File: src/main/java/com/sleepwalker/sleeplib/util/PermissionUtil.java
package com.sleepwalker.sleeplib.util;

import net.minecraft.command.CommandSource;
import net.minecraft.entity.player.ServerPlayerEntity;
import net.minecraftforge.server.permission.PermissionAPI;

import javax.annotation.Nonnull;
import java.util.function.Predicate;

public final class PermissionUtil {

    @Nonnull
    public static Predicate<CommandSource> requires(@Nonnull String node){
        return source -> {
            if(source.getEntity() == null){
                return true;
            }
            else if(source.getEntity() instanceof ServerPlayerEntity){
                return PermissionAPI.hasPermission((ServerPlayerEntity) source.getEntity(), node);
            }
            else return false;
        };
    }
}

File: src/main/kotlin/com/sleepwalker/sleeplib/util/brigadier/PermissionUtil.kt
package com.sleepwalker.sleeplib.util.brigadier

import com.mojang.brigadier.builder.ArgumentBuilder
import net.minecraft.command.CommandSource
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraftforge.server.permission.PermissionAPI

fun <S : CommandSource, T : ArgumentBuilder<S, T>> T.requiresNode(node: String): T = this.requires {
    when(val entity = it.entity){
        null -> true
        is ServerPlayerEntity -> PermissionAPI.hasPermission(entity, node)
        else -> false
    }
}